File: common/code/EPizzas.Common/Configuration.cs
using CommunityToolkit.Diagnostics;
using LanguageExt;
using Microsoft.Extensions.Configuration;

namespace EPizzas.Common;

public static class ConfigurationExtensions
{
    public static string GetValue(this IConfiguration configuration, string key) =>
        configuration.TryGetValue(key)
                     .IfNoneThrow($"Could not find '{key}' in configuration.");

    public static Option<string> TryGetValue(this IConfiguration configuration, string key) =>
        configuration.TryGetSection(key)
                     .BindNullable(section => section.Value);

    public static Option<IConfigurationSection> TryGetSection(this IConfiguration configuration, string key)
    {
        Guard.IsNotNull(configuration);

        var section = configuration.GetSection(key);

        return section.Exists()
                ? Option<IConfigurationSection>.Some(section)
                : Option<IConfigurationSection>.None;
    }
}

File: common/code/common.fsharp/Configuration.fs
[<RequireQualifiedAccess>]
module common.Configuration

open Microsoft.Extensions.Configuration

let private getSection (configuration: IConfiguration) (key: string) =
    let section = configuration.GetSection(key)
    if section.Exists() then Some section else None

let getValue configuration key =
    getSection configuration key
    |> Option.bind (fun section ->
        match section.Value with
        | Null -> None
        | NonNull value -> Some value)

let getValueOrThrow configuration key =
    getValue configuration key
    |> Option.defaultWith (fun () -> failwithf $"Could not find value for key '{key}'.")
